dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/IconMapper.java

```diff
@@ -35,71 +35,78 @@
 import org.hisp.dhis.fileresource.FileResource;
 import org.hisp.dhis.fileresource.FileResourceService;
 import org.hisp.dhis.icon.CustomIcon;
-import org.hisp.dhis.icon.DefaultIcon;
-import org.hisp.dhis.icon.Icon;
-import org.hisp.dhis.icon.IconResponse;
-import org.hisp.dhis.schema.descriptors.IconSchemaDescriptor;
-import org.hisp.dhis.user.CurrentUserUtil;
-import org.hisp.dhis.webapi.service.ContextService;
 import org.springframework.stereotype.Component;
 
 @Component
 @AllArgsConstructor
 public class IconMapper {
   private FileResourceService fileResourceService;
 
-  private ContextService contextService;
-
-  public IconResponse from(Icon icon) {
-    if (icon instanceof CustomIcon ci) {
-      return new IconResponse(
-          icon.getKey(),
-          icon.getDescription(),
-          icon.getKeywords(),
-          ci.getFileResourceUid(),
-          ci.getCreatedByUserUid(),
-          getCustomIconReference(ci.getKey()),
-          icon.getCreated(),
-          icon.getLastUpdated());
-    } else {
-      return new IconResponse(
-          icon.getKey(),
-          icon.getDescription(),
-          icon.getKeywords(),
-          getDefaultIconReference(icon.getKey()),
-          icon.getCreated(),
-          icon.getLastUpdated());
+  public CustomIcon to(CustomIconRequest customIconRequest) throws BadRequestException {
+
+    Optional<FileResource> fileResource = Optional.empty();
+
+    if (customIconRequest.getCustom()) {
+
+      if (customIconRequest.getFileResourceId() == null) {
+        throw new BadRequestException("FileResource must be provided with CustomIcon");
+      }
+
+      fileResource =
+          fileResourceService.getFileResource(customIconRequest.getFileResourceId(), CUSTOM_ICON);
+      if (fileResource.isEmpty()) {
+        throw new BadRequestException(
+            String.format(
+                "FileResource with uid %s does not exist", customIconRequest.getFileResourceId()));
+      }
     }
+
+    CustomIcon customIcon = new CustomIcon();
+    customIcon.setKey(customIconRequest.getKey());
+    customIcon.setDescription(customIconRequest.getDescription());
+    customIcon.setKeywords(customIconRequest.getKeywords());
+    customIcon.setCustom(customIconRequest.getCustom());
+    customIcon.setCode(customIconRequest.getCode());
+    customIcon.setFileResource(fileResource.orElse(null));
```

Comment:
Should this ever be null?

Reply:
yes. if there is a change in default icons. no?

Reply:
can default icons be changed by a user via our API?

Reply:
description or addition/removal from keywords. These are cases where default icons can be updated.